src/igraph_eulerian.c

```diff
@@ -0,0 +1,499 @@
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+
+    igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS);
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) % 2 == 1) odd++;
+    }
+
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (odd > 2) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    /* if pdd count is 2, then semi-eulerian
+    if odd count is 0, then eulerian
+    odd count never is 1 for undirected graph 
+    */
+
+    if (odd == 2) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 0;
+        *has_cycle = 1;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+int is_eulerian_directed(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_bool_t res_strong, res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess;
+    igraph_integer_t i;
+    igraph_vector_t out_degree, in_degree;
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS);
+
+    IGRAPH_CHECK(igraph_vector_init(&in_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &in_degree);
+    igraph_degree(graph, &in_degree, igraph_vss_all(), IGRAPH_IN, IGRAPH_LOOPS);
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+            } else if ((VECTOR(out_degree)[i] + 1 == VECTOR(in_degree)[i]) && (incoming_excess < 1 && outgoing_excess < 2)) {
+                incoming_excess++;
+            } else {
+                *has_path = 0;
+                *has_cycle = 0;
+                igraph_vector_destroy(&in_degree);
+                igraph_vector_destroy(&out_degree);
+                return IGRAPH_SUCCESS;
+            }
+        }
+    }
+
+    igraph_vector_destroy(&in_degree);
+    igraph_vector_destroy(&out_degree);
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong)) {
+        *has_path = 0;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong || res_weak)) {
+        *has_path = 1;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    *has_path = 0;
+    *has_cycle = 0;
+
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/* flexible function */
```

Comment:
What does this comment refer to? Remember, very likely people who don't know you will read and modify this code in the future. They won't be able to ask you. The comments are addressed to them.